File: app/src/main/java/com/cn29/aac/ui/feedentry/FeedEntryFragment.java
package com.cn29.aac.ui.feedentry;


import static com.cn29.aac.ui.feedentry.FeedEntryFragment.Mode.LIST;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import com.cn29.aac.R;
import com.cn29.aac.databinding.FragmentFeedentryListBinding;
import com.cn29.aac.repo.feedentry.FeedEntry;
import com.cn29.aac.ui.base.BaseFragment;
import com.cn29.aac.ui.common.BaseRecyclerViewAdapter.OnItemClickListener;
import com.cn29.aac.ui.feedentry.FeedEntryAdapter.MyMenuItemClickListener;
import com.cn29.aac.ui.feedentry.vm.FeedEntryListViewModel;
import com.cn29.aac.ui.feedentrydetail.FeedEntryDetailActivity;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import javax.inject.Inject;


/**
 * A fragment representing a list of Items.
 * <p />
 * Activities containing this fragment MUST implement the {@link }
 * interface.
 */
public class FeedEntryFragment extends BaseFragment {

  @Inject
  FeedEntryListViewModel viewModel;

  @Inject
  FragmentFeedentryListBinding binding;

  Mode mode = LIST;

  private FeedEntryAdapter adapter;

  public void setMode(Mode mode) {
    this.mode = mode;
  }

  @Override
  public void onActivityCreated(@Nullable Bundle savedInstanceState) {
    super.onActivityCreated(savedInstanceState);
    viewModel.getCompositeEntrys().observe(this, entries -> {
      //update the data
      if (adapter == null) {
        OnItemClickListener<FeedEntry> feedEntryOnItemClickListener = getFeedEntryOnItemClickListener();
        MyMenuItemClickListener menuItemClickListener = getMyMenuItemClickListener();
        adapter = new FeedEntryAdapter(entries,
            feedEntryOnItemClickListener, menuItemClickListener);
        binding.list.setAdapter(adapter);
      } else {
        adapter.notifyDataSetChanged();
      }
    });
  }

  @NonNull
  private MyMenuItemClickListener getMyMenuItemClickListener() {
    return new MyMenuItemClickListener() {
      @Override
      public boolean onMenuItemClick(MenuItem item, FeedEntry feedEntry) {
        int menuId = item.getItemId();
        if (menuId == R.id.action_entry_delete) {
          new AlertDialog.Builder(binding.getRoot().getContext())
              .setTitle("Warning").setMessage("Are you sure to delete the Feed Entry?")
              .setPositiveButton("OK",
                  (dialogInterface, j) ->
                      Single.create(subscriber -> viewModel.delete(feedEntry))
                          .subscribeOn(Schedulers.newThread())
                          .observeOn(AndroidSchedulers.mainThread())
                          .subscribe()
              )
              .setNegativeButton("Cancel", (dialogInterface, i) -> dialogInterface.dismiss())
              .create()
              .show();
        } else if (menuId == R.id.action_entry_favourite) {
          feedEntry.setFavourite(!feedEntry.isFavourite());
          Single.create(subscriber -> viewModel.update(feedEntry))
              .subscribeOn(Schedulers.newThread())
              .observeOn(AndroidSchedulers.mainThread())
              .subscribe();
        }
        return false;
      }
    };
  }

  @NonNull
  private OnItemClickListener<FeedEntry> getFeedEntryOnItemClickListener() {
    return item -> {
      Intent intent = new Intent(getActivity(), FeedEntryDetailActivity.class);
      intent.putExtra(FeedEntryDetailActivity.EXTRA_POSITION, item.getUid());
      startActivity(intent);
    };
  }


  @Override
  public View onCreateView(LayoutInflater inflater, ViewGroup container,
      Bundle savedInstanceState) {
    // Set the adapter
    switch (mode) {
      case GRID:
        binding.list.setLayoutManager(new GridLayoutManager(getActivity(), 4));
        break;
      case LIST:
        binding.list.setLayoutManager(new LinearLayoutManager(getActivity()));
        break;
      case TILE:
        binding.list.setLayoutManager(new GridLayoutManager(getActivity(), 2));
        break;
    }
    return binding.getRoot();
  }

  public enum Mode {
    LIST, GRID, TILE
  }

}

File: app/src/main/java/com/cn29/aac/ui/feedentry/FeedEntryFragment.kt
package com.cn29.aac.ui.feedentry

import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AlertDialog
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import com.cn29.aac.R
import com.cn29.aac.databinding.FragmentFeedentryListBinding
import com.cn29.aac.repo.feedentry.FeedEntry
import com.cn29.aac.ui.base.BaseFragment
import com.cn29.aac.ui.common.BaseRecyclerViewAdapter
import com.cn29.aac.ui.feedentry.FeedEntryAdapter.MyMenuItemClickListener
import com.cn29.aac.ui.feedentry.vm.FeedEntryListViewModel
import com.cn29.aac.ui.feedentry.vm.FeedEntryListViewModel.CompositeModel
import com.cn29.aac.ui.feedentrydetail.FeedEntryDetailActivity
import io.reactivex.Single
import io.reactivex.SingleEmitter
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject

/**
 * A fragment representing a list of Items.
 *
 *
 * Activities containing this fragment MUST implement the []
 * interface.
 */
class FeedEntryFragment : BaseFragment() {
    @JvmField
    @Inject
    var viewModel: FeedEntryListViewModel? = null

    @JvmField
    @Inject
    var binding: FragmentFeedentryListBinding? = null
    private var mode = Mode.LIST
    private var adapter: FeedEntryAdapter? = null
    fun setMode(mode: Mode) {
        this.mode = mode
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        viewModel!!.compositeEntrys.observe(viewLifecycleOwner,
                                            Observer { entries: CompositeModel? ->
                                                //update the data
                                                if (adapter == null) {
                                                    val feedEntryOnItemClickListener = feedEntryOnItemClickListener
                                                    val menuItemClickListener = myMenuItemClickListener
                                                    adapter = FeedEntryAdapter(
                                                            entries!!,
                                                            feedEntryOnItemClickListener,
                                                            menuItemClickListener)
                                                    binding!!.list.adapter = adapter
                                                } else {
                                                    adapter!!.notifyDataSetChanged()
                                                }
                                            })
    }

    private val myMenuItemClickListener: MyMenuItemClickListener
        get() = object : MyMenuItemClickListener() {
            override fun onMenuItemClick(item: MenuItem?,
                                         feedEntry: FeedEntry?): Boolean {
                val menuId = item!!.itemId
                if (menuId == R.id.action_entry_delete) {
                    AlertDialog.Builder(binding!!.root
                                                .context)
                            .setTitle("Warning")
                            .setMessage("Are you sure to delete the Feed Entry?")
                            .setPositiveButton("OK"
                            ) { _: DialogInterface?, _: Int ->
                                Single.create { _: SingleEmitter<Any?>? ->
                                            viewModel!!.delete(
                                                    feedEntry)
                                        }
                                        .subscribeOn(
                                                Schedulers.newThread())
                                        .observeOn(
                                                AndroidSchedulers.mainThread())
                                        .subscribe()
                            }
                            .setNegativeButton("Cancel"
                            ) { dialogInterface: DialogInterface, _: Int -> dialogInterface.dismiss() }
                            .create()
                            .show()
                } else if (menuId == R.id.action_entry_favourite) {
                    feedEntry!!.isFavourite = !feedEntry.isFavourite
                    Single.create { _: SingleEmitter<Any?>? ->
                                viewModel!!.update(feedEntry)
                            }
                            .subscribeOn(Schedulers.newThread())
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe()
                }
                return false
            }
        }

    private val feedEntryOnItemClickListener: BaseRecyclerViewAdapter.OnItemClickListener<FeedEntry?>
        get() = object :
                BaseRecyclerViewAdapter.OnItemClickListener<FeedEntry?> {
            override fun onItemClick(item: FeedEntry?) {

                val intent = Intent(
                        activity,
                        FeedEntryDetailActivity::class.java)
                intent.putExtra(
                        FeedEntryDetailActivity.EXTRA_POSITION,
                        item!!.uid)
                startActivity(
                        intent)
            }
        }

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Set the adapter
        when (mode) {
            Mode.GRID -> binding!!.list.layoutManager = GridLayoutManager(
                    activity, 4)
            Mode.LIST -> binding!!.list.layoutManager = LinearLayoutManager(
                    activity)
            Mode.TILE -> binding!!.list.layoutManager = GridLayoutManager(
                    activity, 2)
        }
        return binding!!.root
    }

    enum class Mode {
        LIST, GRID, TILE
    }
}